### PR TITLE
bpo-41197: Async magic methods in contextlib.closing

### PR DESCRIPTION
**Pull request by uburuntu**
# Async magic methods in contextlib.closing

I was surprised that `contextlib` has no async analogue of this `closing` class, because async scripts often use any kind of closings and I think `__aenter__` and `__aexit__` methods should be added to `contextlib.closing`, so that we can use `contextlib.closing` in async code too. 

For example:

```python3
class SomeAPI:
    ...

    async def request(self):
        pass

    async def aclose(self):
        await self.session.close()
        

async with closing(SomeAPI()) as api:
    response = await api.request()
    print(response)
```

Also these methods can be moved to another class (like `asyncclosing` along the lines of `asynccontextmanager`).

https://bugs.python.org/issue41197

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@uburuntu

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

File: Lib/test/test_contextlib_async.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import asyncio
-from contextlib import asynccontextmanager, AbstractAsyncContextManager, AsyncExitStack
+from contextlib import asynccontextmanager, closing, AbstractAsyncContextManager, AsyncExitStack
```
Comment (ZackerySpytz) [inline Lib/test/test_contextlib_async.py]: Please limit modified lines to 79 characters to follow PEP 8.

File: Misc/NEWS.d/next/Library/2020-07-02-22-00-21.bpo-41197.zE6sX-.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Added `__aenter__` and `__aexit__` methods to :class:`contextlib.closing`,
```
Comment (ZackerySpytz) [inline Misc/NEWS.d/next/Library/2020-07-02-22-00-21.bpo-41197.zE6sX-.rst]:
```suggestion
Added ``__aenter__`` and ``__aexit__`` methods to :class:`contextlib.closing`,
```
Please use double backquotes around `__aenter__` and `__aexit__`.

File: Lib/contextlib.py [hunk 3]
```diff
@@ -294,13 +294,29 @@ class closing(AbstractContextManager):
...
     def __init__(self, thing):
         self.thing = thing
     def __enter__(self):
         return self.thing
     def __exit__(self, *exc_info):
         self.thing.close()
+    async def __aenter__(self):
+        return self.thing
+    async def __aexit__(self, *exc_info):
+        await self.thing.close()
```
Comment (1st1) [inline Lib/contextlib.py]: Should be `aclose()`

**Review by 1st1 — CHANGES_REQUESTED** [review]
+1, but the closing method should be `aclose()`

**Comment by uburuntu** [comment]
Thanks for the review.

I have made the requested changes; please review again.

**Comment by njsmith** [comment]
I think this and #21545 accidentally raced with each other to solve the same problem, and #21545 won the race; its `aclosing` helper will be shipped in 3.10 next week. (Also having a separate `aclosing` does seem slightly better than overloading `closing` to call two different methods depending on context, IMO.) So unfortunately there may not be anything left to do with this except, well... closing it. It sucks when this kind of thing happens, but not much to be done :-(

**Comment by uburuntu** [comment]
@njsmith, thanks for the clarification.

It's depressing, but that's okay. Good luck 🚀

**Pull request closed without merging**